app/data_model/database.py

```diff
@@ -68,8 +71,32 @@ def create_session_and_engine():
         logger.exception(e)
         raise e
 
-
 db_session, engine = create_session_and_engine()
 base.query = db_session.query_property()
 
 base.metadata.create_all(engine)
+
+
+@event.listens_for(engine, "engine_connect")
+def test_connection(connection, branch):
+    if branch:
+        # "branch" refers to a sub-connection of a connection,
+        # we don't want to bother testing on these.
+        return
```

Comment:
It's a listener function, it doesn't return anything. The idea being it allows you to execute some functionality when a connection is checked out of the pool. In this case we test the connection, if it fails tries to reconnect and if the reconnect fails it then forcibly closes the connection.

It's a modified version of:
http://docs.sqlalchemy.org/en/latest/core/pooling.html#disconnect-handling-pessimistic